DEVDOC.md

```diff
@@ -396,3 +396,265 @@ bazel-bin/gapii/cc/vulkan_spy_1.cpp
+
+This is oriented to AGI developers who need to work with AGI internals. It is
+meant to help build a high-level mental model that facilitates navigating the
+actual source code. It references concepts and code modules that are unlikely to
+change anytime soon, and it tries not to reference details that are likely to
+change. You are encouraged to check the source code while reading this, to
+clarify how the various parts actually fits together. If you spot any
+discrepancy, please update this doc, but please keep it high-level.
+
+AGI is architectured to be API-agnostic in order to support multiple graphics
+API. GAPID, from which AGI was forked, supports GLES and Vulkan. AGI main focus
+is Vulkan on Android, so all examples in this doc are related to Vulkan on
+Android.
+
+### AGI overall architecture
+
+AGI is separated in main components that interact mostly via protobuf:
+
+- GAPIS (Graphics API Server, Go code under `gapis/`, see `gapis/README.md`) is
+  the main component, running on the developer desktop/laptop. Any complex logic
+  is meant to be implemented in GAPIS, while the other components are meant to
+  be kept as simple as possible. The protobuf interface of GAPIS is defined in
+  `gapis/service/service.proto`.
+
+- GAPII (Graphics API Interceptor, mostly C++ code under `gapii/`, see
+  `gapii/README.md`) is the component responsible for intercepting graphics API
+  calls during capture.  It interacts with GAPIS via a dedicated protocol.
+
+- GAPIR (Graphics API Replayer, mostly C++ code under `gapir/`, see
+  `gapir/README.md`) is the component that can replay graphics API calls. Its
+  protobuf interface is defined in `gapir/replay_service/service.proto`.
+
+- GAPIC (Graphics API Client, Java code under `gapic/`, see `gapic/README.md`)
+  is the GUI client. It uses GAPIS protobuf interface.
+
+- GAPIT (Graphics API Terminal, Go code under `cmd/gapit/`) is the
+  developer-oriented CLI client. It uses GAPIS protobuf interface.
+
+When you start AGI via the desktop icon or the `./agi` command, this triggers
+the entry point defined under `cmd/agi/`, which by default starts a new GAPIS
+and then a new GAPIC that connects to this GAPIS. GAPIS may then itself start
+GAPII or GAPIR instances.
+
+To capture and replay on Android, AGI has APKs (one per ABI, e.g.
+`gapid-arm64-v8a.apk`) that embeds GAPII and GAPIR.
+
+### Create a gfxtrace: capture, serialize and store Vulkan calls
+
+AGI uses a Vulkan layer to intercept and capture the Vulkan calls emitted by an
+application, and store them into a gfxtrace file. Be aware that in the code
+base, the act of intercepting graphics API calls may be referred by the words
+"capture", "trace", "intercept" and "spy".
+
+For a Vulkan capture on Android, the main steps are:
+
+1. GAPIS does a few adb commands to edit global settings that tell the Android
+   Vulkan loader to inserts AGI's Vulkan layer when starting the app to capture.
+   This layer is called `GraphicsSpy`, it is implemented by
```

Comment:
This layer is called `GraphicsSpy` and is implemented by

Reply:
done